pyccl/nl_pt/ept.py

```diff
@@ -0,0 +1,697 @@
+            bias terms in the expansion. `'linear'`: use the linear
+            matter power spectrum. `'nonlinear'`: use the non-linear
+            matter power spectrum. `'pt'`: use the 1-loop SPT matter
+            power spectrum. Default: `'nonlinear'`.
+        bk2_pk_kind (str): power spectrum to use for the non-local
+            bias terms in the expansion. Same options and default as
+            `b1_pk_kind`.
+        pad_factor (float): fraction of the log10(k) interval you
+             to add as padding for FFTLog calculations. Default: 1.0.
+        low_extrap (float): decimal logaritm of the minimum Fourier
+             scale (in Mpc^-1) for which FAST-PT will extrapolate.
+             Default: -5.0.
+        high_extrap (float): decimal logaritm of the maximum Fourier
+             scale (in Mpc^-1) for which FAST-PT will extrapolate.
+             Default: 3.0.
+        P_window (array_like): 2-element array describing the
+             tapering window used by FAST-PT. See FAST-PT
+             documentation for more details. Default: `None`.
+        C_window (float):  `C_window` parameter used by FAST-PT to
+             smooth the edges and avoid ringing. See FAST-PT
+             documentation for more details. Default: 0.75.
+        sub_lowk (bool): if `True`, the small-scale white noise
+             contribution to some of the terms will be subtracted.
+             Default: `False`.
+    """
+    __repr_attrs__ = ('with_NC', 'with_IA', 'with_matter_1loop',
+                      'k_s', 'a_s', 'exp_cutoff',
+                      'b1_pk_kind', 'bk2_pk_kind', 'fastpt_par')
+
+    def __init__(self, *, with_NC=False, with_IA=False,
+                 with_matter_1loop=True, cosmo=None,
+                 log10k_min=-4, log10k_max=2, nk_per_decade=20,
+                 a_arr=None, k_cutoff=None, n_exp_cutoff=4,
+                 b1_pk_kind='nonlinear', bk2_pk_kind='nonlinear',
+                 pad_factor=1.0, low_extrap=-5.0, high_extrap=3.0,
+                 P_window=None, C_window=0.75, sub_lowk=False):
+        self.with_matter_1loop = with_matter_1loop
+        self.with_NC = with_NC
+        self.with_IA = with_IA
+
+        # Set FAST-PT parameters
+        self.fastpt_par = {'pad_factor': pad_factor,
+                           'low_extrap': low_extrap,
+                           'high_extrap': high_extrap,
+                           'P_window': P_window,
+                           'C_window': C_window,
+                           'sub_lowk': sub_lowk}
+
+        to_do = ['one_loop_dd']
+        if self.with_NC:
+            to_do.append('dd_bias')
+        if self.with_IA:
+            to_do.append('IA')
+
+        # k sampling
+        nk_total = int((log10k_max - log10k_min) * nk_per_decade)
+        self.k_s = np.logspace(log10k_min, log10k_max, nk_total)
+
+        # a sampling
+        if a_arr is None:
+            a_arr = get_pk_spline_a()
+        self.a_s = a_arr.copy()
```

Comment:
Why are you copying it? There is no need. You are producing a new array inside the function, and even when the user passes one, you are not modifying it. Copying takes time.

Reply:
I'm pretty sure this copy is needed. Check below
```
>>> class My(object):
...     def __init__(self, a):
...             self.a = a
... 
>>> a = np.arange(10)
>>> a
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
>>> m = My(a)
>>> 
>>> m.a
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
>>> a[0] = 3
>>> a
array([3, 1, 2, 3, 4, 5, 6, 7, 8, 9])
>>> m.a
array([3, 1, 2, 3, 4, 5, 6, 7, 8, 9])
```

Reply:
I.e. if the initial array is modified outside the object, it will also be modified inside, since it gets assigned by reference.

Reply:
You are correct. I didn't think of the chance of the array mutating outside of the object. Resolving.